x-pack/plugins/observability_solution/observability_onboarding/server/routes/kubernetes/route.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { v4 as uuidv4 } from 'uuid';
+import * as t from 'io-ts';
+import Boom from '@hapi/boom';
+import { termQuery } from '@kbn/observability-plugin/server';
+import type { estypes } from '@elastic/elasticsearch';
+import { getFallbackESUrl } from '../../lib/get_fallback_urls';
+import { createObservabilityOnboardingServerRoute } from '../create_observability_onboarding_server_route';
+import { hasLogMonitoringPrivileges } from '../../lib/api_key/has_log_monitoring_privileges';
+import { createShipperApiKey } from '../../lib/api_key/create_shipper_api_key';
+
+interface CreateKubernetesOnboardingFlowRouteResponse {
+  apiKeyEncoded: string;
+  onboardingId: string;
+  elasticsearchUrl: string;
+}
+
+interface HasKubernetesDataRouteResponse {
+  hasData: boolean;
+}
+
+const createKubernetesOnboardingFlowRoute = createObservabilityOnboardingServerRoute({
+  endpoint: 'POST /internal/observability_onboarding/kubernetes/flow',
+  options: { tags: [] },
+  async handler({
+    context,
+    request,
+    plugins,
+    services,
+  }): Promise<CreateKubernetesOnboardingFlowRouteResponse> {
+    const {
+      elasticsearch: { client },
+    } = await context.core;
+
+    const hasPrivileges = await hasLogMonitoringPrivileges(client.asCurrentUser);
+
+    if (!hasPrivileges) {
+      throw Boom.forbidden(
+        "You don't have enough privileges to start a new onboarding flow. Contact your system administrator to grant you the required privileges."
+      );
+    }
+
+    const fleetStart = await plugins.fleet.start();
+    const packageClient = fleetStart.packageService.asScoped(request);
+
+    await packageClient.ensureInstalledPackage({ pkgName: 'kubernetes' });
+
+    const { encoded: apiKeyEncoded } = await createShipperApiKey(
+      client.asCurrentUser,
+      'onboarding_kubernetes'
+    );
+    const elasticsearchUrlList = plugins.cloud?.setup?.elasticsearchUrl
+      ? [plugins.cloud?.setup?.elasticsearchUrl]
+      : await getFallbackESUrl(services.esLegacyConfigService);
+
+    return {
+      onboardingId: uuidv4(),
+      apiKeyEncoded,
+      elasticsearchUrl: elasticsearchUrlList.length > 0 ? elasticsearchUrlList[0] : '',
+    };
+  },
+});
+
+const hasKubernetesDataRoute = createObservabilityOnboardingServerRoute({
+  endpoint: 'GET /internal/observability_onboarding/kubernetes/{onboardingId}/has-data',
```

Comment:
Right, we need to check for a different field, even though other parts of the query stay the same. It seemed not worth it to generalize over `esClient.search` even further, though I understand there is a bit of duplication. My thinking is that this generalized endpoint might over time become just a huge `switch` statement to handle each use-case to check for data.

Reply:
I'm happy with that approach. The best abstraction will emerge the more of these different types of integrations we cater for and we can tweak these endpoints going forwards since they're all marked as internal.